resource/unarchive/unarchive.go

```diff
@@ -0,0 +1,383 @@
+	}
+
+	status.AddDifference("unarchive", u.Source, u.Destination, "")
+	status.RaiseLevelForDiffs()
+
+	return nil
+}
+
+// setDirsAndContents sets the Unarchive fields of unarchive destination and its
+// contents, and the temporary fetch/unarchive destination and its contents. A
+// bool indicating whether duplicates need to be evaluated between the unarchive
+// destination and the temporary fetch/unarchive destination.
+func (u *Unarchive) setDirsAndContents() (bool, error) {
+	var err error
+
+	// set the unarchive destination directory
+	u.destDir, err = os.Open(u.Destination)
+	if err != nil {
+		return false, err
+	}
+
+	// walk the destination directory to set the destination contents
+	filepath.Walk(u.destDir.Name(), func(path string, f os.FileInfo, err error) error {
+		u.destContents = append(u.destContents, path)
+		return nil
+	})
+
+	// read the contents of the temporary fetch/unarchive location
+	fetchDir := u.fetchLoc
+	u.fetchDir, err = os.Open(fetchDir)
+	if err != nil {
+		return false, err
+	}
+	fetchDirContents, err := u.fetchDir.Readdir(0)
+	if err != nil {
+		return false, errors.Wrapf(err, "could not read dir %q", fetchDir)
+	}
+
+	// if one directory is within the temporary fetch/unarchive location, we need
+	// to use this as the directory to read file names
+	if len(fetchDirContents) == 1 && fetchDirContents[0].IsDir() {
+		fetchDir = u.fetchLoc + "/" + fetchDirContents[0].Name()
+		u.fetchDir, err = os.Open(fetchDir)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// walk the fetch directory to set the fetch contents
+	filepath.Walk(u.fetchDir.Name(), func(path string, f os.FileInfo, err error) error {
+		u.fetchContents = append(u.fetchContents, path)
+		return nil
+	})
+
+	// if there are no files, we do not need to compare checksums with files in
+	// the temporary fetch/unarchive location. We check whether the length is 1
+	// because the directory is in the contents.
+	if len(u.destContents) == 1 {
+		return false, nil
+	}
+	return true, nil
+}
+
+// evaluateDuplicates evaluates whether identical files exist in u.Destination
+// and the temporary fetch/unarchive location
+func (u *Unarchive) evaluateDuplicates() error {
+	// determine which directory has fewer items in order to minimize operations
+	dirA := u.destDir.Name()
+	dirB := u.fetchDir.Name()
+	filesA := u.destContents
+	filesB := u.fetchContents
+	if len(u.fetchContents) < len(u.destContents) {
+		dirA = u.fetchDir.Name()
+		dirB = u.destDir.Name()
+		filesA = u.fetchContents
+		filesB = u.destContents
+	}
+
+	// for each item in filesA, determine if it also exists in filesB
+	// compare the checksums for the files - if mismatch, return an error
+	for _, fA := range filesA {
+		for _, fB := range filesB {
+			fileA := strings.TrimPrefix(fA, dirA)
+			fileB := strings.TrimPrefix(fB, dirB)
+
+			faStat, err := os.Stat(fA)
+			if err != nil {
+				return err
+			}
+			fbStat, err := os.Stat(fB)
+			if err != nil {
+				return err
+			}
+
```

Comment:
Just a thought here, but would it make sense to compare the size of the files before doing a checksum?  It seems like in the most common case the files would differ in size as well as contents, so we could skip the actual time it takes to generate a checksum.